pinot-core/src/main/java/org/apache/pinot/core/operator/filter/H3InclusionIndexFilterOperator.java

```diff
@@ -0,0 +1,128 @@
+import java.util.List;
+import org.apache.pinot.common.request.context.ExpressionContext;
+import org.apache.pinot.common.request.context.ExpressionContext.Type;
+import org.apache.pinot.common.request.context.predicate.Predicate;
+import org.apache.pinot.core.common.Operator;
+import org.apache.pinot.core.operator.blocks.FilterBlock;
+import org.apache.pinot.core.operator.dociditerators.ScanBasedDocIdIterator;
+import org.apache.pinot.core.operator.docidsets.BitmapDocIdSet;
+import org.apache.pinot.segment.local.utils.GeometrySerializer;
+import org.apache.pinot.segment.local.utils.H3Utils;
+import org.apache.pinot.segment.spi.IndexSegment;
+import org.apache.pinot.segment.spi.index.reader.H3IndexReader;
+import org.apache.pinot.spi.utils.BytesUtils;
+import org.locationtech.jts.geom.Geometry;
+import org.roaringbitmap.buffer.BufferFastAggregation;
+import org.roaringbitmap.buffer.ImmutableRoaringBitmap;
+import org.roaringbitmap.buffer.MutableRoaringBitmap;
+
+
+/**
+ * A filter operator that uses H3 index for geospatial data inclusion
+ */
+public class H3InclusionIndexFilterOperator extends BaseFilterOperator {
+
+  private static final String EXPLAIN_NAME = "INCLUSION_FILTER_H3_INDEX";
+
+  private static final String OPERATOR_NAME = "H3InclusionIndexFilterOperator";
+
+  private final IndexSegment _segment;
+  private final Predicate _predicate;
+  private final int _numDocs;
+  private final H3IndexReader _h3IndexReader;
+  private final LongSet _h3Ids;
+
+  public H3InclusionIndexFilterOperator(IndexSegment segment, Predicate predicate, int numDocs) {
+    _segment = segment;
+    _predicate = predicate;
+    _numDocs = numDocs;
+
+    List<ExpressionContext> arguments = predicate.getLhs().getFunction().getArguments();
+    Geometry geometry;
+    // Assume first argument is Literal, and second argument is IDENTIFIER for St_Contains.
+    assert arguments.get(1).getType() == Type.IDENTIFIER;
+    assert arguments.get(0).getType() == Type.LITERAL;
+    // look up arg1's h3 indices
+    _h3IndexReader = segment.getDataSource(arguments.get(1).getIdentifier()).getH3Index();
+    // arg0 is the literal
+    geometry = GeometrySerializer.deserialize(BytesUtils.toBytes(arguments.get(0).getLiteral()));
+    // must be some h3 index
+    assert _h3IndexReader != null;
+
+    // get the set of H3 cells at the specified resolution which completely cover the input shape.
+    _h3Ids = H3Utils.coverGeometryInH3(geometry, _h3IndexReader.getH3IndexResolution().getLowestResolution());
+  }
+
+  @Override
+  protected FilterBlock getNextBlock() {
+    // have list of h3 cell ids for polygon provided
+    // return filtered num_docs
+    ImmutableRoaringBitmap[] partialMatchDocIds = new ImmutableRoaringBitmap[_h3Ids.size()];
+    int i = 0;
+    for (long h3IndexId : _h3Ids) {
+      partialMatchDocIds[i++] = _h3IndexReader.getDocIds(h3IndexId);
+    }
+    MutableRoaringBitmap mutableRoaringBitmap = BufferFastAggregation.or(partialMatchDocIds);
```

Comment:
I think this is just approximation, but it doesn't return same result of `ST_Contains`, right?

Reply:
i think 
```
expressionFilterOperator.getNextBlock().getBlockDocIdSet().iterator();
    MutableRoaringBitmap result = docIdIterator.applyAnd(partialMatchDocIds);
```
will do the final check on the potential match set.

Reply:
okay, it could be further optimized to skip the cells fully contained in the polygon. But we can leave it for futture (add a TODO) if you find it too difficult.

Reply:
sure. will create a TODO.

Reply:
https://github.com/apache/pinot/issues/8547

Reply:
actually I did the optimization. but still will use the issue to track if H3 support different polyfill natively.